server/tracedb/azureappinsights.go

```diff
@@ -115,70 +115,181 @@ func (db *azureAppInsightsDB) GetTraceByID(ctx context.Context, traceID string)
 	return parseAzureAppInsightsTrace(traceID, table)
 }
 
-type columnIndex map[string]int
+type spanTable struct {
+	rows []spanRow
+}
+
+func (st *spanTable) Spans() []spanRow {
+	output := make([]spanRow, 0)
+	for _, row := range st.rows {
+		if row.Type() != "trace" {
+			output = append(output, row)
+		}
+	}
+
+	return output
+}
+
+func (st *spanTable) Events() []spanRow {
+	output := make([]spanRow, 0)
+	for _, row := range st.rows {
+		if row.Type() == "trace" {
+			output = append(output, row)
+		}
+	}
+
+	return output
+}
+
+type spanRow struct {
+	values map[string]any
+}
+
+func (sr *spanRow) Get(name string) any {
+	return sr.values[name]
+}
+
+func (sr *spanRow) Type() string {
+	return sr.values["itemType"].(string)
+}
+
+func (sr *spanRow) ParentID() string {
+	return sr.values["operation_ParentId"].(string)
+}
+
+func (sr *spanRow) SpanID() string {
+	return sr.values["id"].(string)
+}
+
+func newSpanTable(table *azquery.Table) spanTable {
+	spanRows := make([]spanRow, 0, len(table.Rows))
+	for _, row := range table.Rows {
+		spanRows = append(spanRows, newSpanRow(row, table.Columns))
+	}
+
+	return spanTable{spanRows}
+}
+
+func newSpanRow(row azquery.Row, columns []*azquery.Column) spanRow {
+	values := make(map[string]any)
+	for i, column := range columns {
+		name := *column.Name
+		if value := row[i]; value != nil {
+			values[name] = value
+		}
+	}
+
+	return spanRow{values}
+}
 
 func parseAzureAppInsightsTrace(traceID string, table *azquery.Table) (model.Trace, error) {
-	columnIndex := mapColumnNames(table.Columns)
-	spans := make([]model.Span, len(table.Rows))
+	spans, err := parseSpans(table)
+	if err != nil {
+		return model.Trace{}, err
+	}
+
+	return model.NewTrace(traceID, spans), nil
+}
 
-	for i, row := range table.Rows {
-		span, err := parseRowToSpan(row, columnIndex)
+func parseSpans(table *azquery.Table) ([]model.Span, error) {
+	spanTable := newSpanTable(table)
+	spanRows := spanTable.Spans()
+	eventRows := spanTable.Events()
```

Comment:
AppInsights returns everything as `rows`. So we use the field `itemType` to know if that is a span (`itemType != "trace"`) or an event (`itemType == "trace"`).